firebase-firestore/src/main/java/com/google/firebase/firestore/auth/FirebaseAppCheckTokenProvider.java

```diff
@@ -0,0 +1,117 @@
+   * {@link Provider} once the "AppCheck" module becomes available.
+   */
+  @Nullable private InternalAppCheckTokenProvider internalAppCheckTokenProvider;
+
+  /** The AppCheck token string. May be null if the token has not been retrieved yet. */
+  @Nullable private String appCheckToken;
+
+  /** Creates a new FirebaseAppCheckTokenProvider. */
+  public FirebaseAppCheckTokenProvider(
+      Deferred<InternalAppCheckTokenProvider> deferredAppCheckTokenProvider) {
+    if (deferredAppCheckTokenProvider == null) {
+      return;
+    }
+    deferredAppCheckTokenProvider.whenAvailable(
+        provider -> {
+          internalAppCheckTokenProvider = provider.get();
+          if (internalAppCheckTokenProvider == null) {
+            return;
+          }
+          Task<AppCheckTokenResult> pendingResult =
+              internalAppCheckTokenProvider.getToken(/* forceRefresh= */ false);
+          if (pendingResult == null) {
+            return;
+          }
+          pendingResult.addOnCompleteListener(
+              task -> {
+                synchronized (this) {
+                  if (appCheckToken != null) {
+                    return;
+                  }
+                  if (task.isSuccessful()) {
+                    AppCheckTokenResult result = task.getResult();
+                    if (result.getError() != null) {
+                      Logger.warn(
+                          LOG_TAG,
+                          "Error getting App Check token; using placeholder token instead. Error: "
+                              + result.getError());
+                    }
+                    appCheckToken = result.getToken();
+                  } else {
+                    Logger.warn(
+                        LOG_TAG,
+                        "Unexpected error getting App Check token: " + task.getException());
+                  }
+                }
+              });
+        });
+  }
+
+  @Nullable
+  public synchronized String getCurrentAppCheckToken() {
+    if (appCheckToken != null) {
+      return appCheckToken;
+    }
+
+    if (internalAppCheckTokenProvider == null) {
+      return null;
+    }
+
+    try {
+      Task<AppCheckTokenResult> pendingResult =
+          internalAppCheckTokenProvider.getToken(/* forceRefresh= */ false);
+      if (pendingResult == null) {
+        return null;
+      }
+      AppCheckTokenResult result =
+          Tasks.await(pendingResult, MAXIMUM_TOKEN_WAIT_TIME_MS, TimeUnit.MILLISECONDS);
+      if (result.getError() != null) {
+        Logger.warn(
+            LOG_TAG,
+            "Error getting App Check token; using placeholder token instead. Error: "
+                + result.getError());
+      }
+      appCheckToken = result.getToken();
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      Logger.warn(LOG_TAG, "Unexpected error getting App Check token: " + e);
+    }
+    return appCheckToken;
```

Comment:
If I understand this correctly, you are repeating the steps from the constructor here so you can block on the token result. You don't have to do do this - you should create a new instance variable that is just the Task. You can create a new task using [TaskCompletionSource](https://developers.google.com/android/reference/com/google/android/gms/tasks/TaskCompletionSource), which you can then fulfill in the constructor and block on here.

You should also not block in a synchronized method. The scope of your synchronization needs to be much narrower (but note that you won't have to use synchronization at all if you use Tasks).

Reply:
Done. This was very useful. Thanks for your guidance.